### ISSUE TITLE
CRITICAL issue Aspect of the Viper

### ISSUE THREAD

**Issue by mayday2013** [SunstriderEmu/BugTracker#2679]
**Describe the bug**
Hunter's Aspect of the Viper does not regen mana properly when it drops <60%

**To Reproduce**
I have tested mana regen provided by this aspect on a different server.
- Drop your mana <40 points
- Activate AotV(costs 40 mana)
- Stay in combat and watch your mana regenerating with the same speed <60% and >60% of total mana 

Spell: https://wowwiki.fandom.com/wiki/Aspect_of_the_Viper

**Expected behavior**
Regen more mana at <60% of total mana
**Screenshots/videos**
https://youtu.be/aMSrBBsq9Go video test of how your mana regens at 220 int
on 2:18 I got full mana
on endless: https://sun9-33.userapi.com/c206728/v206728497/c86f2/LCHLnVX16B8.jpg
Full mana got on this timer: https://sun9-42.userapi.com/c206728/v206728497/c86fc/v52jJszZpCc.jpg

**Additional context**
You are oom after 2 minutes of intense fight and then you can't properly regen mana for pet ressurection, steady shots etc. Please investigate this issue.

**Comment by Folgrum** [comment]
I have noticed this as well that it barely gives any mana.

**Comment by mchayes** [comment]
Barely any mana, huge amount of threat. Crazy bugged after changelog that was supposed to fix it.

**Comment by mayday2013** [comment]
https://youtu.be/pReB7SErgDA

This is how it works on arena.

**Comment by Lunasoll** [comment]
> https://youtu.be/pReB7SErgDA
> 
> This is how it works on arena.
> 
> Got 2 weps enchanted with 30 int, total of 218 intellect. Unplayable in >2 min fights

you can't compare it with PvP situations. The way you spend mana is totally different in PvP then in PvE.

**Comment by Ironjudgment** [comment]
Hey, so I did some research/math on this, oh boy let's get geeky :D

All of my calculations are based on this video (of myself):
https://youtu.be/nqYwcIWwkT0
https://imgur.com/a/CkqwfZp

So normal mana regeneration for all classes is as follows:
 `MP5 = 5 * (0.001 + sqrt(Int) * Spirit * Base_Regen) rounded up` [1]

I did some calculations and came to conclusion that my base regen should be about ~52 MP5
After observing my mana regen, the actual normal rate was exactly 58 MP5. So slightly higher than calculated.
![image](https://user-images.githubusercontent.com/33090790/78511717-de80bc80-7796-11ea-994b-3ac0f60d5cc8.png)
Link to the spreadsheet: https://docs.google.com/spreadsheets/d/1MTMwImWwydFtMie0dHsnDVBQWOqv3OF_-RBOBrWKJqY/edit?usp=sharing

To calculate the Aspect of the Viper I've found an old wowgamepedia articale from 13th of May 2008 so a couple of months before WotLK at the time of 2.4.3 [2].

Unfortunately the graph showing the coefficients of Viper is no longer there but from the text I have found two solid leads:
`With at least 90% mana, the effect is 11% of total Intellect.`
`This is followed by a linear increase up to 55% of total Intellect until the mana bar is down to 20%`
So, 11% coefficient at or above 90% mana and 55% coefficient at or below 20% mana.

Some calculations again:
![image](https://user-images.githubusercontent.com/33090790/78512772-273c7380-779f-11ea-8c88-b07184f0d78c.png)
Link to the spreadsheet: https://docs.google.com/spreadsheets/d/1MTMwImWwydFtMie0dHsnDVBQWOqv3OF_-RBOBrWKJqY/edit?usp=sharing

I would like to point out that the 'Expected' value is calculated with current mana value for each row.

The results came back a little weird in my understanding, so data shows that <=20% mana MP5 is scaled somewhat correctly, there aren't any BIG inconsistencies but as mana increases the Vipers effects seems to scale less well. Especially at >90% of mana where it regens way too much.

Additionally, I have a few notes I would like to highlight:
1. I do not know if passive regen should stack with Viper? If anyone finds a source for this that would be great!
2. In patch notes from [2]: 
`Patch 2.4.2 (2008-05-13): This ability now grants an additional amount of mana each 5 seconds equal to 35% of the hunter's level. (24.5 mp5 at lvl70)` From this it seems that Viper should regen its formula value + 24.5 MP5 which there isn't any indication in the data that it does.

Please let me know if I made any blunders while calulating, also I would love to hear from the Devs on how viper is scripted :D

[1]https://wow.gamepedia.com/index.php?title=Mana_regeneration&oldid=1393686
[2]https://wow.gamepedia.com/index.php?title=Aspect_of_the_Viper&oldid=1364347

**Comment by mayday2013** [comment]
> > https://youtu.be/pReB7SErgDA
> > This is how it works on arena.
> > Got 2 weps enchanted with 30 int, total of 218 intellect. Unplayable in >2 min fights
> 
> you can't compare it with PvP situations. The way you spend mana in PvP is totally different then in PvE. In PvP, you use a lot more abilities so it's normal to oom even with AotV. However in PvE you're not supposed to be oom after a couple of fights. At the moment on Endless, there is no difference between using AotV or any other aspect in terms of mana generation.

I'm not comparing PvE with PvP. Im showing in-combat mana regerenationm which is stupidly low.

If all of this comes from having 200 int in s1, let me know, probably, we should wait till s2 for better mana regen as hunters have 350+ int in s2 gear.

**Comment by mayday2013** [comment]
Ironjudgement, greetings! thanks for your job done on this question!

Did you take into consideration patchnote 2.3.2 which states the follows:
 Patch 2.3.2 (2008-01-08): Effect increased.

And yes, mana gained from AotV is an additional mana you regenerate to your usual mp5 rate.

What I'm trying to say here is that it makes me confused that hunters are that hihgly mana-dependant at this moment. I've playing hunters for a while on different pservers, usually starting full s1/s2 gear and I've never had such situations in PvP screnarios.

Let me know if I'm mistaken and AotV works here fine.

**Comment by Ironjudgment** [comment]
Hey @mayday2013 the formula I've used to calculate Viper efficiency at <=20% of mana does use 0.55.
To answer your second question: I don't know, is the short answer.

If we take what you said as true
`And yes, mana gained from AotV is an additional mana you regenerate to your usual mp5 rate.`
then: Viper(+24.5MP5) + passive MP5 should be hunters regen while Viper is active. From data we can see that at lower levels of mana, the amount of mana that is regenerated corresponds to the Viper forumla **without the additional 24.5 MP5**. BUT, at higher levels of mana(>=90%)the effect of Viper is still too high.

Taking my previously calculated values,
![image](https://user-images.githubusercontent.com/33090790/78538089-65aa5080-77e8-11ea-9217-a704258774f5.png)

Conclusion:
Viper is performing better at higher % of mana and is underperforming while at lower % of mana.

**Comment by Lunasoll** [comment]
So I had a SH(N) run on my lvl 66 hunter in AotV. It wasn't actually that bad. I could spam steady shot macro without really going oom. Granted I did have BoW from pala.

**Comment by Aeded** [comment]
@Ironjudgment You are making it more complicated to yourself and to everyone whos reading it. Base regen MP5 ticks every 2 seconds (its the small ticks of 29 mana in the vid). So you can completely disregard those and just focus on the "bigger" mana-gains every 5s that are obviously from AotV and just base the calculation on those.

Thats why your column "M. Gained" does not make much sense, because sometimes you wrote down small MP5 tick with AotV combined. 
Therefore making a "Difference" and "Avarage" column does not make sense at all either.

**Comment by Ironjudgment** [comment]
@Aeded I do not see your point there. I measured mana gained every 5 secs which would include every possible mana gained in that time and compared it to the Viper calculation. 

While doing this I found out that every second mp5 gain is always larger by 27 and when you even them out with previous smaller gain they exactly even out.

The columns, Average and Expected correspond to the average mana gained per 5 secs compared to the expected amount from Viper.

Additionally, Viper is missing 24.5MP5 effect increase and most likely the base mp5 regen is not there either if you compare mana gained and expected value from JUST viper alone.

Eseentially I measure the overall mana gained and break it down to see if the values correspond to expected values. How else was I supposed to check if Viper is working correctly? I am open to suggestions and different results :]

**Comment by Aeded** [comment]
@Ironjudgment Viper aspect has hidden aura that ticks every 5s and it is completely independant on base MP5. Therefore I made the comment explaining, that if you are making report about Aspect of Viper it is much easier to read an more understandable for both parties if you include only numbers relevant to the AotV MP5. 

I did not say you made an incorrect conclusion **based on the sources you worked with**, I just poited out that you chose the least efficient method.

What would I personaly do, would be making data set of _(Current_Mana / Max_Mana)_ with lots of values, ranging from 1 to 5608 on _(Current_Mana)_. For the each fraction of % I would calculate the exact MP5 **expected** from Aspect of Viper, while having variable **INT**. 

After that, I would look at the video you made and marked down only the "bigger" MP5 gains from AotV. I would note down mana before the tick, mana after the tick, get from it the **actual** MP5 gain from AotV and finaly compare it to the data of **expected** MP5.

Now at this point, you would see that the actual MP5 and expected MP5 (based on WowWiki) do not fit and it is very easy to read & understand, because it is only about AotV mana regeneration, and base MP5 is not mixed in it.

However, with a natural bit of skepticism and armed with dynamic sheet you start to test out different values of **INT** that hunters will have in future. Slowly you start to realize that the numbers go quite steeply through the roof on **expected** MP5 from AotV, and you gain a suspicion, which leads you to one and only trustworthy source to prove anything - retail videos...

And from there on starts a way to the bitter end for hunters mana, wowwiki and your report.

PS: If you are not familliar on how much INT hunters have around seasons 3 & 4 it is between 290-340. With 340 being the BiS gear.
I also recommend searching this very bugtracker for "viper" and you can see all the reports made on this issue.

**Comment by Ironjudgment** [comment]
@Aedad I didn't make up any numbers in this. I've always used my stats in all calculations, you can view all formulas in the spreadsheet.

> What would I personally do, would be making data set of (Current_Mana / Max_Mana) with lots of values, ranging from 1 to 5608 on (Current_Mana). For each fraction of % I would calculate the exact MP5 expected from Aspect of Viper, while having variable INT

I did exactly that.
It's my fault for not labelling the columns more appropriately and explaining my process more thoroughly.

The approach I took was to find the current mana and record current mana again in 5 secs. since MP5 ticks every ~2.5secs I recorded mana in 4s and 6s increments (total 10s). Then compare mana gained in the increments of 5 to the expected Viper regen.

In detail:
At 0:28 secs current mana is 167 then at 0:32 its 338 then at 0:38 it's 536.

Now, the total mana gained from 0:28-0:32 is 171 and total mana gained from 0:32-0:38 is 198. Now, there is always a 27 mana difference between these two MP5 cycles so I just averaged them which comes to 184.5 MP5.

Note that this is total mana gained, which includes all regen effects.

What I did next was to calculate how much mana should AotV regen at that point. From wowwiki we know that <=20% the coefficient is 0.55 so following the viper formula:

MP5Viper = Intellect × 22/35 × ( 0.9 - Manacurrent / Manamax ) + Intellect × 0.55
167 * 22/35 * ( 0.9 - 167 / 5608 ) + 167 * 0.55 = 183.198
Viper alone should have regen ~183.198 mana.
Note that at that time total mana regened was 171.

For the next cycle of MP5:
MP5Viper = Intellect × 22/35 × ( 0.9 - Manacurrent / Manamax ) + Intellect × 0.55
167 * 22/35 * ( 0.9 - 338 / 5608 ) + 167 * 0.55 = 179.998

So in these 2 MP5 cycles(10 secs) total mana regened by Viper should have been 363.196 where as actual total mana regened was 369.

From these values I've come to the conclusion that the increased viper effect (24.5 additional MP5) is not included and the base MP5 regen is also not there, but at high % of mana, the total mana regen is too high even if you include all mana regen effects together.

**Comment by Aeded** [comment]
> The approach I took was to find the current mana and record current mana again in 5 secs. since MP5 ticks every ~2.5secs I recorded mana in 4s and 6s increments (total 10s). Then compare mana gained in the increments of 5 to the expected Viper regen.

yes I do understand what you did, just saying that its complicated for no reason... By the way all MP5 ticks every 2 second exactly, not ~2.5secs. Value of each MP5 tick is _2***MP5**/5_

I would have to repeat myself again, you clearly dont understand what I tried to tell you and I got a feeling we would be going back and forth about it for a long time. 
So I will do you a favor and get straight to the point, the wowwiki formula is wrong and sadly the formula they are using right now is correct.

If you dont believe me, look at any retail PvP video from 2.4+ check how much mana they are gaining, then test the wowwiki formula for the amount of intellect they usualy have in that gear (290-340)

**Comment by Skuzzlez** [comment]
Can we get an update on this? This issue is huge and honestly gamebreaking for hunters in PvP right now.

**Comment by Enory** [comment]
Heres are my values on my hunt.
![image](https://user-images.githubusercontent.com/25368657/82664453-0a1a1200-9c32-11ea-8bf7-96a6d372284d.png)

Full video of regen here were i destacked passive regen (every 2.5sec) and aotv regen (every 5sec on the timer).
https://youtu.be/RbmJXyb-OHA

**Comment by Skuzzlez** [comment]
Update:

Using this PvP video from 2008 - you can clearly see the hunter's mana ticks from Aspect of the Viper.

> https://www.youtube.com/watch?v=4jbSgaPROH0

I looked at the exact mana values that the hunter was receiving from Aspect of the Viper at various percentages of mana and compared them to the values we would expect from the current formula used on Endless, and the WoW Wiki formula provided. This hunter's data is at the bottom of my spreadsheet, Ironjudgment's is at the top:

> https://docs.google.com/spreadsheets/d/1Wg5NpHk4OPyXC7rZyLcO6xqPWEnOnEYi69r5qoVoGgo/edit#gid=0



In order to get his mana values roughly in line with the current formula on Endless, this hunter would've needed approximately 338 to 340 Intellect - which seems accurate for S4 gear. Using the WoWWiki formula, this hunter would've received vastly different amounts of mana.

The amount of mana the hunter is receiving is exactly in line with endless, except under certain situations: When he is below 20% mana, and when he is above 90% mana. This is corroborated by the fact that he receives the exact same mana (211) when he is at 17% mana, and at 20% mana, Similarly, he receives roughly the exact same mana at 93% mana as he does at 97% mana (61 vs 62) therefore somewhat proving that the aforementioned caps are real:

> With at least 90% mana, the effect is 11% of total Intellect

> This is followed by a linear increase up to 55% of total Intellect until the mana bar is down to 20%

TLDR:

The endless formula is mostly correct - it has the total and minimum values returned correct, but does not have the caps correctly implemented. The maximum mana returned by Aspect of the Viper (based on the 55% total intellect coefficient) should be reached when the hunter is at 20% mana, and not at 0%. Similarly, the minimum amount returned (based on the 11% total intellect coefficient) should be reached when the hunter is at 90% mana, and all values above that. The curve should be adjusted accordingly.

**Comment by kelno** [comment]
Thanks for all that research here.
So after research what the server is currently doing for the gain from intelligence part is slightly incorrect.
The formula for this part currently is:
intellectPercent = 10 + (100 - powerPercent) * 0.45
Which indeed doesn't take into accounts the 20% to 90% caps. So for example a player at 20% mana receives 80% of the benefit instead of 100%. I'll fix that now.

**Comment by kelno** [comment]
Applied with next server update

**Comment by kelno** [comment]
It's live now
